File: mars/src/main/java/MoveForward.java
/**
 * @author guang
 * @since 2019-08-22 14:20
 */
public class MoveForward extends Move {
    public MoveForward(Area area) {
        super(area);
    }

    @Override
    public Point execute(Point point) {
        switch (point.getDirection()) {
            case SOUTH:
                point.setY(point.getY() + 1);
                break;
            case NORTH:
                point.setY(point.getY() - 1);
                break;
            case EAST:
                point.setX(point.getX() + 1);
                break;
            case WEST:
                point.setX(point.getX() - 1);
            default:
        }

        // 是否超出区域
        if (isCrossing(point)) {
            throw new RuntimeException("Rover 超出探索区域" + point.toString());
        }

        return point;
    }
}

File: mars-java-kotlin/src/main/kotlin/MoveForward.kt
class MoveForward(area: Area) : Move(area) {

    override fun excute(point: Point): Point {

        when (point.direction) {
            Direction.SOUTH -> point.y = point.y + 1
            Direction.NORTH -> point.y = point.y - 1
            Direction.WEST -> point.x = point.x - 1
            Direction.EAST -> point.x = point.x + 1
        }

        if (isCrossing(point)) {
            throw RuntimeException("越移动界point : $point")
        }

        return point
    }
}